token/services/network/fabric/lookup/channelflm.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package lookup
+
+import (
+	"context"
+	"encoding/base64"
+	"sync"
+	"time"
+
+	"github.com/hyperledger-labs/fabric-smart-client/pkg/utils/errors"
+	"github.com/hyperledger-labs/fabric-smart-client/platform/fabric"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/network/common/rws/translator"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap/zapcore"
+)
+
+type ChannelListenerManagerConfig struct {
+	MaxRetries        int
+	RetryWaitDuration time.Duration
+}
+
+type channelBasedFLMProvider struct {
+	fnsp           *fabric.NetworkServiceProvider
+	tracerProvider trace.TracerProvider
+	keyTranslator  translator.KeyTranslator
+	config         ChannelListenerManagerConfig
+}
+
+func NewChannelBasedFLMProvider(fnsp *fabric.NetworkServiceProvider, tracerProvider trace.TracerProvider, keyTranslator translator.KeyTranslator, config ChannelListenerManagerConfig) *channelBasedFLMProvider {
+	return &channelBasedFLMProvider{
+		fnsp:           fnsp,
+		tracerProvider: tracerProvider,
+		keyTranslator:  keyTranslator,
+		config:         config,
+	}
+}
+
+func (p *channelBasedFLMProvider) NewManager(network, channel string) (ListenerManager, error) {
+	net, err := p.fnsp.FabricNetworkService(network)
+	if err != nil {
+		return nil, err
+	}
+	ch, err := net.Channel(channel)
+	if err != nil {
+		return nil, err
+	}
+	return &channelBasedLLM{
+		network:   network,
+		channel:   ch,
+		listeners: make(map[string]*Scanner),
+	}, nil
+}
+
+type channelBasedLLM struct {
+	network string
+	channel *fabric.Channel
+
+	listenersMutex sync.RWMutex
+	listeners      map[string]*Scanner
+}
+
+func (c *channelBasedLLM) AddLookupListener(namespace string, key string, startingTxID string, stopOnLastTx bool, listener Listener) error {
+	s := &Scanner{
+		context:      nil,
+		channel:      c.channel,
+		namespace:    namespace,
+		key:          key,
+		startingTxID: startingTxID,
+		stopOnLastTx: stopOnLastTx,
+		listener:     listener,
+	}
+	c.listenersMutex.Lock()
+	c.listeners[key] = s
+	c.listenersMutex.Unlock()
+	go s.Scan()
+	return nil
+}
+
+func (c *channelBasedLLM) RemoveLookupListener(id string, listener Listener) error {
+	c.listenersMutex.Lock()
+	defer c.listenersMutex.Unlock()
+
+	s, ok := c.listeners[id]
+	if ok {
+		s.Stop()
+		delete(c.listeners, id)
+	}
+	return nil
+}
+
+type Scanner struct {
+	context      context.Context
+	channel      *fabric.Channel
+	namespace    string
+	key          string
```

Comment:
it might make sense to accept multiple keys here so that we can batch the scans and not launch one scan per key.

Reply:
okay